MoreLinq.Test/TakeLastTest.cs

```diff
@@ -73,11 +57,29 @@ public void TakeLastDisposesSequenceEnumerator()
             {
                 seq.TakeLast(1).Consume();
             }
+        }
 
-            using (var seq = new[]{ 1,2,3 }.AsTestingSequence())
-            {
-                seq.TakeLast(1).Consume();
-            }
+        // ICollection<T> Optimization Tests
```

Comment:
These tests don't test for _optimisation_ so the comment is misleading. They only test that the _behaviour_ of `TakeLast` does not change when passed a collection, which is why removing the optimisation in `TakeLast` does not fail them.

It seems futile to test that `TakeLast` is optimised for collections. The best one can do is to assert that the `Count` property was used, which doesn't seem like a very useful assurance. My main problem with 4ce583392328975ea5a5d197e893fdf829654a5f was that it seemed to abuse local functions when a simpler and more classic approach to have a (locally) reusable assertion function would be sufficient, like this:

```c#
static void AssertTakeLast<T>(IEnumerable<T> input, int count, IEnumerable<T> expected)
{
    // TODO comment why we are testing twice, once via an identity projection
    input.Select(x => x).TakeLast(count).AssertSequenceEqual(expected);
    input.TakeLast(count).AssertSequenceEqual(expected);
}

public void TakeLast()
{
    AssertTakeLast(new[] { 12, 34, 56, 78, 910, 1112 }, 3, new[] { 78, 910, 1112 });
}
```

f68966e54af292acfd2819583832c4694b669675 seems to add a lot of duplication of test data and code with little clarity.

Reply:
> how can I make a test that fail if the optimisation is removed?

Any suggestion?